File: modules/rng.py
import discord
from discord.ext import commands
import random
import string

class FunGamesCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command(name='coinflip')
    async def coin_flip(self, ctx):
        result = random.choices(['Heads', 'Tails', 'Edge'], weights=[49.5, 49.5, 1], k=1)[0]
        await ctx.send(f"The coin landed on: {result}")

    @commands.command(name='diceroll')
    async def dice_roll(self, ctx, sides: int):
        if sides not in [6, 20]:
            await ctx.send("Please choose a 6-sided or 20-sided dice.")
            return
        result = random.randint(1, sides)
        await ctx.send(f"You rolled a {sides}-sided dice and got: {result}")

    @commands.command(name='randomstring')
    async def random_string(self, ctx, length: int):
        if length <= 0 or length > 100:
            await ctx.send("Please specify a length between 1 and 100.")
            return
        result = ''.join(random.choices(string.ascii_letters + string.digits, k=length))
        await ctx.send(f"Your random string: {result}")

async def setup(bot):
    await bot.add_cog(FunGamesCog(bot))

File: modules/rng.js
const { SlashCommandBuilder } = require('discord.js');
const { customId, parseCustomId, buttons } = require('../lib/interactions');

const MODULE = 'rng';
const rerollRow = (type, max) => buttons([{ id: customId(MODULE, 'reroll', `${type}:${max ?? ''}`), label: 'Losuj ponownie', style: 'secondary', emoji: '🔄' }]);

function roll(type, max) {
  if (type === 'coinflip') {
    return `🪙 ${Math.random() < 0.495 ? 'Heads' : Math.random() < 0.99 ? 'Tails' : 'Edge'}`;
  }
  if (type === 'diceroll') {
    const sides = [6, 20].includes(max) ? max : 6;
    return `🎲 ${Math.floor(Math.random() * sides) + 1} (d${sides})`;
  }
  if (type === 'number') {
    const cap = max || 100;
    return `🔢 ${Math.floor(Math.random() * cap) + 1}`;
  }
  if (type === 'randomstring') {
    const len = Math.min(Math.max(max || 16, 1), 128);
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let s = '';
    for (let i = 0; i < len; i++) s += chars[Math.floor(Math.random() * chars.length)];
    return `\`${s}\``;
  }
  return null;
}

module.exports = {
  data: new SlashCommandBuilder()
    .setName('rng')
    .setDescription('Random generators')
    .addStringOption(option =>
      option.setName('type')
        .setDescription('Type of generator')
        .setRequired(true)
        .setAutocomplete(true)
    )
    .addIntegerOption(option =>
      option.setName('max')
        .setDescription('Max / length / sides')
        .setRequired(false)
    ),
  async autocomplete(interaction) {
    const focused = interaction.options.getFocused();
    const choices = ['coinflip', 'diceroll', 'number', 'randomstring'];
    const filtered = choices.filter(choice => choice.startsWith(focused));
    await interaction.respond(filtered.map(choice => ({ name: choice, value: choice })));
  },
  async execute(interaction) {
    const type = interaction.options.getString('type');
    const max = interaction.options.getInteger('max');
    const result = roll(type, max);
    if (result == null) {
      await interaction.reply('Unknown type');
      return;
    }
    await interaction.reply({ content: result, components: rerollRow(type, max) });
  },

  // Central component router (main.js) dispatches here for any customId
  // prefixed "rng:" — see lib/interactions.js and INTERACTIONS.md.
  async handleComponent(interaction) {
    const { payload } = parseCustomId(interaction.customId);
    const [type, maxStr] = payload.split(':');
    const max = maxStr ? Number(maxStr) : undefined;
    const result = roll(type, max);
    await interaction.update({ content: result ?? 'Unknown type', components: result != null ? rerollRow(type, max) : [] });
  }
};
